mautic/mautic
# Citrix oAuth2 refactoring

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR adds unit tests to the code submitted in https://github.com/mautic/mautic/pull/7380.
Also it contains refactoring of the code from the previously mentioned PR.

But with 7380, the refresh token was not taken into account so the plugin had to be manually re-authenticated every hour. This PR fixes that by adding the necessary support from the legacy integration framework to refresh access tokens using a refresh token.

#### Steps to reproduce the bug:
1. Set up one of the Goto integrations, authorize then wait 1 hour
1. Create a campaign to add the contact to a Goto event (meeting, webinar, etc)
1. Notice that the drop down list is empty and the log shows "token expired" error
1. Steps to test this PR:
1. Load up this PR
1. Repeat and the access token will refresh using the refresh token

Code hack to simulate:
1. It helps to have xdebug setup or else you'll have to extract the unencrypted access token to confirm that it changed
1. Add a xdebug breakpoint at https://github.com/mautic/mautic/blob/07c2c5a5edb3576304684afa2f80eea97fc343e7/app/bundles/PluginBundle/Integration/AbstractIntegration.php#L1499.
1. Before that line add $this->keys['expires'] = 1;
1. Add a campaign action from the step 2 of the steps to reproduce bug above
1. Note the value in $this->keys['access_token'] through xdebug
1. Repeat step 4 and note that the access token is now different
